lib: Add hash function for nexthop groups

sworleys opened the pull request:
Add a hash function to turn a nexthop group into a
32 bit unsigned hash key with jhash. We do not care to
hash any recursively resolved nexthops, just the group.

Signed-off-by: Stephen Worley <sworley@cumulusnetworks.com>

louberger commented on lib/nexthop_group.c:
```diff
@@ -147,6 +148,42 @@ void copy_nexthops(struct nexthop **tnh, struct nexthop *nh,
+	 */
+	for (nh = nhg->nexthop; nh; nh = nh->next) {
+		key = jhash_1word(nh->type, key);
+		key = jhash_2words(nh->vrf_id, nh->nh_label_type, key);
```
Why not jhash_3words instead of 1 followed by 2words?

louberger commented on lib/nexthop_group.c:
```diff
@@ -147,6 +148,42 @@ void copy_nexthops(struct nexthop **tnh, struct nexthop *nh,
+		key = jhash(&nh->rmap_src, sizeof(nh->rmap_src), key);
+		if (nh->nh_label) {
+			for (i = 0; i < nh->nh_label->num_labels; i++)
+				key = jhash_1word(nh->nh_label->label[i], key);
```
I think it's worth special casing the 2 and 3word case, i.e., something like
i = nh->nh_label->num_labels;
while (i >= 3) {
   key = jhash_3words(nh->nh_label->label[i], nh->nh_label->label[i+1],nh->nh_label->label[i+2],key);
   i -=3;
}
if (i >= 2) {
   key = jhash_2words(nh->nh_label->label[i], nh->nh_label->label[i+1],key);
   i -=2;
}
if (i >= 2) {
   key = jhash_1word(nh->nh_label->label[i], key);
   i -=2;
}

sworleys commented:
> This are suggested changes - if you really don't want to make them, I won't object.

@louberger I implemented your suggested changes and also moved the hash function directly into lib/nexthop.c per @mjstapp 's suggestion offline.

mjstapp commented on lib/nexthop.c:
```diff
@@ -316,12 +316,47 @@ unsigned int nexthop_level(struct nexthop *nexthop)
+
+		while (i >= 3) {
+			key = jhash_3words(nexthop->nh_label->label[i],
+					   nexthop->nh_label->label[i + 1],
```
we can't really use `i` this way, can we - aren't we looking beyond the end of the array? I think we need two variables - one counting forward from zero, and a second one counting down the remaining number of labels to process.

sworleys replied:
Oh man you are 100% right, sorry I didn't even notice that in the suggestion.

mjstapp replied:
looks good now!

mjstapp commented on lib/nexthop.c:
```diff
@@ -316,12 +316,50 @@ unsigned int nexthop_level(struct nexthop *nexthop)
 
 uint32_t nexthop_hash(struct nexthop *nexthop)
```
could `nexthop` be `const`?

mjstapp commented on lib/nexthop_group.c:
```diff
@@ -147,6 +148,21 @@ void copy_nexthops(struct nexthop **tnh, const struct nexthop *nh,
 	}
 }
 
+uint32_t nexthop_group_hash(struct nexthop_group *nhg)
```
could `nhg` be `const` ?

louberger commented:
Yes, sorry bad pseudo code on my part...


----------
On April 5, 2019 9:40:02 AM Mark Stapp <notifications@github.com> wrote:

> mjstapp requested changes on this pull request.
>
> Thanks for moving the logic into the individual nexthop - but I had a 
> concern about the label array indexing.
>
>>  
> +	if (nexthop->nh_label) {
> +		int i = nexthop->nh_label->num_labels;
> +
> +		while (i >= 3) {
> +			key = jhash_3words(nexthop->nh_label->label[i],
> +					   nexthop->nh_label->label[i + 1],
>
> we can't really use `i` this way, can we - aren't we looking beyond the end 
> of the array? I think we need two variables - one counting forward from 
> zero, and a second one counting down the remaining number of labels to process.
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/FRRouting/frr/pull/4066#pullrequestreview-223279186
